File: lab2/source/del.cpp
#include "operations.h"
#include "factory.h"

class Del : public Operations{
public:
    void Operation(std::stack<int> &stack, std::istream& fin) override {
        Empty_check(stack);
        int previous_top = stack.top();
        stack.pop();
        Empty_check(stack);
        if(previous_top == 0){
            throw std::runtime_error("Error: division by zero");
        }
        int res =  stack.top() / previous_top;
        stack.pop();
        stack.push(res);
    }
};

namespace {
    Operations* createDel() {
        return new Del;
    }
    bool x = OperationsFactory::getInstance()->RegisterOperation(std::string( "/"), createDel);
}

File: jaba/lab1/src/main/java/org/example/Del.java
package org.example;

import java.io.OutputStream;
import java.util.Scanner;
import java.util.Stack;

public class Del extends Operations{
    @Override
    public void Operation(Stack<Integer> stack, Scanner scanner, OutputStream s){
        emptyCheck(stack);
        int previous_top = stack.pop();
        emptyCheck(stack);
        if (stack.peek() == 0) {
            throw new DevByZeroException("Error: division by zero");
        }
        int res =  stack.pop() / previous_top;
        stack.push(res);
    }
}
